contracts/proxy-lib/src/lib.rs

```diff
@@ -0,0 +1,408 @@
+pub struct ProxyContract {
+    pub context_id: ContextId,
+    pub context_config_account_id: AccountId,
+    pub num_approvals: u32,
+    pub proposal_nonce: ProposalId,
+    pub proposals: IterableMap<ProposalId, Proposal>,
+    pub approvals: IterableMap<ProposalId, HashSet<SignerId>>,
+    pub num_proposals_pk: IterableMap<SignerId, u32>,
+    pub active_proposals_limit: u32,
+    pub context_storage: IterableMap<Box<[u8]>, Box<[u8]>>,
+}
+
+#[derive(Clone, Debug)]
+#[near(serializers = [borsh])]
+pub struct FunctionCallPermission {
+    allowance: Option<U128>,
+    receiver_id: AccountId,
+    method_names: Vec<String>,
+}
+
+#[derive(Clone, PartialEq)]
+#[near(serializers = [json, borsh])]
+pub struct ProposalApprovalWithSigner {
+    pub proposal_id: ProposalId,
+    pub signer_id: Repr<SignerId>,
+    pub added_timestamp: u64,
+}
+
+#[derive(Clone, PartialEq, Debug)]
+#[near(serializers = [json, borsh])]
+pub enum ProposalAction {
+    ExternalFunctionCall {
+        receiver_id: AccountId,
+        method_name: String,
+        args: Base64VecU8,
+        deposit: NearToken,
+        gas: Gas,
+    },
+    Transfer {
+        receiver_id: AccountId,
+        amount: NearToken,
+    },
+    SetNumApprovals {
+        num_approvals: u32,
+    },
+    SetActiveProposalsLimit {
+        active_proposals_limit: u32,
+    },
+    SetContextValue {
+        key: Box<[u8]>,
+        value: Box<[u8]>,
+    },
+}
+
+// The proposal the user makes specifying the receiving account and actions they want to execute (1 tx)
+#[derive(Clone, PartialEq, Debug)]
+#[near(serializers = [json, borsh])]
+pub struct Proposal {
+    pub author_id: Repr<SignerId>,
+    pub actions: Vec<ProposalAction>,
+}
+
+#[near]
+impl ProxyContract {
+    #[init]
+    pub fn init(context_id: Repr<ContextId>, context_config_account_id: AccountId) -> Self {
+        Self {
+            context_id: context_id.rt().expect("Invalid context id"),
+            context_config_account_id,
+            proposal_nonce: 0,
+            proposals: IterableMap::new(b"r".to_vec()),
+            approvals: IterableMap::new(b"c".to_vec()),
+            num_proposals_pk: IterableMap::new(b"k".to_vec()),
+            num_approvals: 3,
+            active_proposals_limit: 10,
+            context_storage: IterableMap::new(b"l"),
+        }
+    }
+
+    pub fn create_and_approve_proposal(&self, proposal: Signed<Proposal>) -> Promise {
+        // Verify the signature corresponds to the signer_id
+        let proposal = proposal
+            .parse(|i| *i.author_id)
+            .expect("failed to parse input");
+
+        let author_id = proposal.author_id;
+
+        let num_proposals = self.num_proposals_pk.get(&author_id).unwrap_or(&0) + 1;
+        assert!(
+            num_proposals <= self.active_proposals_limit,
+            "Account has too many active proposals. Confirm or delete some."
+        );
+        return self.perform_action_by_member(MemberAction::Create {
```

Comment:
Hm, they can take advantage of this only being processed in the next block to submit multiple proposals at once, hoping they'll be included in the same block, at which point they'll subvert this limitation

we should increment this number before the config check

Reply:
will do later